### PR TITLE
[Behat][Promotions] Promotions with fixed discount scenarios

### PR DESCRIPTION
**Pull request by Zales0123**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Fixed tickets |
| License       | MIT
| Doc PR        |

First portion of new scenarios for *Promotion*, describing simple fixed discount on customer's cart.
It uses "@todo" tag introduced by @pamil in https://github.com/Sylius/Sylius/pull/4169.

### CODE REVIEW

File: features/shop/apply_fixed_discount_promotion_on_order.feature [hunk 1]
```diff
@@ -0,0 +1,34 @@
+@todo
```
Comment (pamil) [inline features/shop/apply_fixed_discount_promotion_on_order.feature]: :+1: for adding this tag in a new line

File: features/shop/apply_fixed_discount_promotion_on_order.feature [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    I want to have promotions applied
+
+    Background:
+        Given the store is operating on a single channel
+        And default currency is "EUR"
+        And the store has a product "PHP T-Shirt" priced at "€100.00"
+        And the store has a product "PHP Mug" priced at "€6.00"
+        And the store has a product "Symfony Mug" priced at "€10.00"
+        And there is user "john@example.com" identified by "password123"
+        And I am logged in as "john@example.com"
```
Comment (pamil) [inline features/shop/apply_fixed_discount_promotion_on_order.feature]: I guess we already have a step for logging in as customer?

Comment (Zales0123) [inline features/shop/apply_fixed_discount_promotion_on_order.feature]: Not yet in master ;/ https://github.com/Sylius/Sylius/pull/4112

File: features/shop/apply_fixed_discount_promotion_on_order.feature [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+    Scenario: Applying proper promotion discount
+        Given there is a promotion "Total based promotion"
+        And it gives "€10.00" fixed discount for customers with carts above "€80.00"
+        And I have product "PHP T-Shirt" in the cart
+        Then my cart total should be "€90.00"
+        And my cart promotions should be "-€10.00"
+
+    Scenario: Reset cart total if promotion discount is equal with items total
+        Given there is a promotion "Item based promotion"
```
Comment (pjedrzejewski) [inline features/shop/apply_fixed_discount_promotion_on_order.feature]: We should give a better example here, "Christmas Sale" is ok.

File: features/shop/apply_fixed_discount_promotion_on_order.feature (hunk 1 above)
Comment (michalmarcinkowski) [inline features/shop/apply_fixed_discount_promotion_on_order.feature]: Add @todo tag to each scenario not the whole feature ;)

File: features/shop/apply_fixed_discount_promotion_on_order.feature (hunk 2 above)
Comment (michalmarcinkowski) [inline features/shop/apply_fixed_discount_promotion_on_order.feature]: Do I need to be logged in at all?

File: features/shop/apply_fixed_discount_promotion_on_order.feature [hunk 4]
```diff
@@ -0,0 +1,37 @@
+@todo
+@ui-cart
+Feature: Apply fixed discount promotion on order
+    In order to pay proper amount while buying promoted goods
+    As a Customer
+    I want to have promotions applied
```
Comment (michalmarcinkowski) [inline features/shop/apply_fixed_discount_promotion_on_order.feature]: applied **to my cart**

File: features/shop/apply_fixed_discount_promotion_on_order.feature (hunk 2 above)
Comment (Zales0123) [inline features/shop/apply_fixed_discount_promotion_on_order.feature]: Good catch, there is no need of being logged in ;)

File: features/shop/apply_fixed_discount_promotion_on_order.feature [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    Scenario: Applying proper promotion discount
+        Given there is a promotion "Total based promotion"
+        And it gives "€10.00" fixed discount for customers with carts above "€80.00"
+        And I have product "PHP T-Shirt" in the cart
+        Then my cart total should be "€90.00"
+        And my cart promotions should be "-€10.00"
+
+    Scenario: Reset cart total if promotion discount is equal with items total
+        Given there is a promotion "Item based promotion"
+        And it gives "€16.00" fixed discount for customers with carts with more than 1 item
```
Comment (michalmarcinkowski) [inline features/shop/apply_fixed_discount_promotion_on_order.feature]: for customers ~~with carts~~ with more than 1 item **in the cart**

File: features/shop/apply_fixed_discount_promotion_on_order.feature [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+        Then my cart total should be "€0.00"
+        And my cart promotions should be "-€16.00"
+
+    @todo
+    Scenario: Set cart total to €0.00 if promotion discount is bigger than items total
+        Given there is a promotion "Thanksgiving sale"
+        And it gives "€10.00" fixed discount for customers with carts above "€5.00"
+        And I have product "PHP Mug" in the cart
+        Then my cart total should be "€0.00"
+        And my cart promotions should be "-€5.00"
```
Comment (pamil) [inline features/shop/apply_fixed_discount_promotion_on_order.feature]: PHP Mug price is `€6.00`, so cart promotions should be `-€6.00` too.

Comment (Zales0123) [inline features/shop/apply_fixed_discount_promotion_on_order.feature]: :+1:

File: features/shop/apply_fixed_discount_promotion_on_order.feature [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+        And default currency is "EUR"
+        And the store has a product "PHP T-Shirt" priced at "€100.00"
+        And the store has a product "PHP Mug" priced at "€6.00"
+        And the store has a product "Symfony Mug" priced at "€10.00"
+
+    @todo
+    Scenario: Applying proper promotion discount
+        Given there is a promotion "Holiday promotiong"
+        And it gives "€10.00" fixed discount for customers with carts above "€80.00"
+        And I have product "PHP T-Shirt" in the cart
```
Comment (michalmarcinkowski) [inline features/shop/apply_fixed_discount_promotion_on_order.feature]: When

Comment (Zales0123) [inline features/shop/apply_fixed_discount_promotion_on_order.feature]: If ``When``, then it should be ``When I add...`` I suppose?

File: features/shop/apply_fixed_discount_promotion_on_order.feature [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+        Then my cart total should be "€90.00"
+        And my cart promotions should be "-€10.00"
+
+    @todo
+    Scenario: Reset cart total if promotion discount is equal with items total
+        Given there is a promotion "Christmas Sale"
+        And it gives "€16.00" fixed discount for customers with more than 1 item in the cart
+        And I have product "PHP Mug" in the cart
+        And I have product "Symfony Mug" in the cart
```
Comment (michalmarcinkowski) [inline features/shop/apply_fixed_discount_promotion_on_order.feature]: When I add
And I add

File: features/shop/apply_fixed_discount_promotion_on_order.feature [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+        Then my cart total should be "€0.00"
+        And my cart promotions should be "-€16.00"
+
+    @todo
+    Scenario: Set cart total to €0.00 if promotion discount is bigger than items total
+        Given there is a promotion "Thanksgiving sale"
+        And it gives "€10.00" fixed discount for customers with carts above "€5.00"
+        And I have product "PHP Mug" in the cart
```
Comment (michalmarcinkowski) [inline features/shop/apply_fixed_discount_promotion_on_order.feature]: When I add

File: features/shop/apply_fixed_discount_promotion_on_order.feature [hunk 10]
```diff
@@ -0,0 +1,36 @@
+@ui-cart
+Feature: Apply fixed discount promotion on order
+    In order to pay proper amount while buying promoted goods
+    As a Customer
+    I want to have promotions applied to my cart
+
+    Background:
+        Given the store is operating on a single channel
+        And default currency is "EUR"
```
Comment (michalmarcinkowski) [inline features/shop/apply_fixed_discount_promotion_on_order.feature]: Given the store operates on a single channel in "France"

File: features/shop/apply_fixed_discount_promotion_on_order.feature [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+    Scenario: Reset cart total if promotion discount is equal with items total
+        Given there is a promotion "Christmas Sale"
+        And it gives "€106.00" fixed discount to every order
+        When I add product "PHP T-Shirt" to the cart
+        And I add product "PHP Mug" to the cart
+        Then my cart total should be "€0.00"
+        And my cart promotions should be "-€16.00"
+
+    @todo
+    Scenario: Set cart total to €0.00 if promotion discount is bigger than items total
```
Comment (michalmarcinkowski) [inline features/shop/apply_fixed_discount_promotion_on_order.feature]: Apply discount equal to items total when promotion discount is bigger than items total

File: features/shop/apply_fixed_discount_promotion_on_order.feature [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+        And my cart promotions should be "-€10.00"
+
+    @todo
+    Scenario: Reset cart total if promotion discount is equal with items total
+        Given there is a promotion "Christmas Sale"
+        And it gives "€106.00" fixed discount to every order
+        When I add product "PHP T-Shirt" to the cart
+        And I add product "PHP Mug" to the cart
+        Then my cart total should be "€0.00"
+        And my cart promotions should be "-€16.00"
```
Comment (michalmarcinkowski) [inline features/shop/apply_fixed_discount_promotion_on_order.feature]: -€106

File: features/shop/apply_fixed_discount_promotion_on_order.feature [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+    @todo
+    Scenario: Applying proper promotion discount
+        Given there is a promotion "Holiday promotion"
+        And it gives "€10.00" fixed discount to every order
+        When I add product "PHP T-Shirt" to the cart
+        Then my cart total should be "€90.00"
+        And my cart promotions should be "-€10.00"
+
+    @todo
+    Scenario: Reset cart total if promotion discount is equal with items total
```
Comment (michalmarcinkowski) [inline features/shop/apply_fixed_discount_promotion_on_order.feature]: Apply proper discount when promotion discount is equal to items total

**Comment by michalmarcinkowski** [comment]
I think we should add also the following scenario to this feature:
```
Scenario: The fixed discount does not affect a shipping fee
    Given the store has "DHL" shipping method with "€10.00" fee
    And there is a promotion "Holiday promotion"
    And it gives "€10.00" fixed discount to every order
    And I am logged in customer
    When I add product "PHP T-Shirt" to the cart
    And I proceed selecting "DHL" shipping method
    Then my cart total should be "€100.00"
    And my cart shipping fee should be "€10.00"
    And my cart promotions should be "-€10.00"
```

File: features/shop/apply_fixed_discount_promotion_on_order.feature [hunk 14]
```diff
@@ -0,0 +1,47 @@
+@ui-cart
```
Comment (lchrusciel) [inline features/shop/apply_fixed_discount_promotion_on_order.feature]: `@cart`

File: features/shop/apply_fixed_discount_promotion_on_order.feature [hunk 15]
```diff
@@ -0,0 +1,47 @@
+@cart
+Feature: Apply fixed discount promotion on order
+    In order to pay proper amount while buying promoted goods
+    As a Customer
```
Comment (michalmarcinkowski) [inline features/shop/apply_fixed_discount_promotion_on_order.feature]: As a Visitor

File: features/shop/apply_fixed_discount_promotion_on_order.feature [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+    I want to have promotions applied to my cart
+
+    Background:
+        Given the store is operating on a single channel in "France"
+        And the store has a product "PHP T-Shirt" priced at "€100.00"
+        And the store has a product "PHP Mug" priced at "€6.00"
+
+    @todo
+    Scenario: Applying proper promotion discount
```
Comment (michalmarcinkowski) [inline features/shop/apply_fixed_discount_promotion_on_order.feature]: Receiving fixed discount for my cart

File: features/shop/apply_fixed_discount_promotion_on_order.feature [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+        Given there is a promotion "Holiday promotion"
+        And it gives "€10.00" fixed discount to every order
+        When I add product "PHP T-Shirt" to the cart
+        Then my cart total should be "€90.00"
+        And my cart promotions should be "-€10.00"
+
+    @todo
+    Scenario: Apply proper discount when promotion discount is equal to items total
```
Comment (michalmarcinkowski) [inline features/shop/apply_fixed_discount_promotion_on_order.feature]: Receiving discount equal to the items total of my cart

File: features/shop/apply_fixed_discount_promotion_on_order.feature [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+        Given there is a promotion "Christmas Sale"
+        And it gives "€106.00" fixed discount to every order
+        When I add product "PHP T-Shirt" to the cart
+        And I add product "PHP Mug" to the cart
+        Then my cart total should be "€0.00"
+        And my cart promotions should be "-€106.00"
+
+    @todo
+    Scenario: Apply discount equal to items total when promotion discount is bigger than items total
```
Comment (michalmarcinkowski) [inline features/shop/apply_fixed_discount_promotion_on_order.feature]: Receiving discount equal to the items total of my cart even if the discount is bigger than the items total

File: features/shop/apply_fixed_discount_promotion_on_order.feature [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+        Given there is a promotion "Thanksgiving sale"
+        And it gives "200.00" fixed discount to every order
+        When I add product "PHP Mug" to the cart
+        Then my cart total should be "€0.00"
+        And my cart promotions should be "-€100.00"
+
+    @todo
+    Scenario: The fixed discount does not affect a shipping fee
```
Comment (michalmarcinkowski) [inline features/shop/apply_fixed_discount_promotion_on_order.feature]: Receiving fixed discount does not affect the shipping fee

**Comment by michalmarcinkowski** [comment]
:+1:

**Pull request merged**